indexer/osm_editor.cpp

```diff
@@ -375,30 +375,68 @@ Editor::SaveResult Editor::SaveEditedFeature(EditableMapObject const & emo)
 {
   FeatureID const & fid = emo.GetID();
   FeatureTypeInfo fti;
-  bool const isCreated = IsCreatedFeature(fid);
-  if (isCreated)
+
+  auto const featureStatus = GetFeatureStatus(fid.m_mwmId, fid.m_index);
+  bool const wasCreatedByUser = IsCreatedFeature(fid);
+  if (wasCreatedByUser && featureStatus == FeatureStatus::Untouched)
   {
     fti.m_status = FeatureStatus::Created;
     fti.m_feature.ReplaceBy(emo);
   }
   else
   {
-    fti.m_feature = *m_getOriginalFeatureFn(fid);
+    ASSERT_NOT_EQUAL(featureStatus, FeatureStatus::Deleted, ("Unexpected feature status"));
+
+    fti.m_feature = featureStatus == FeatureStatus::Untouched
+        ? *m_getOriginalFeatureFn(fid)
+        : m_features[fid.m_mwmId][fid.m_index].m_feature;
     fti.m_feature.ReplaceBy(emo);
-    if (AreFeaturesEqualButStreet(fti.m_feature, *m_getOriginalFeatureFn(fid)) &&
-        m_getOriginalFeatureStreetFn(fti.m_feature) == emo.GetStreet())
+    bool const sameAsInMWM = featureStatus != FeatureStatus::Created &&
+        AreFeaturesEqualButStreet(fti.m_feature, *m_getOriginalFeatureFn(fid)) &&
+        m_getOriginalFeatureStreetFn(fti.m_feature) == emo.GetStreet();
+
+    if (featureStatus != FeatureStatus::Untouched)
+    {
+      // A feature was modified and equals to one in editor.
+      auto const editedFeatureInfo = m_features[fid.m_mwmId][fid.m_index];
+      if (AreFeaturesEqualButStreet(fti.m_feature, editedFeatureInfo.m_feature) &&
+          editedFeatureInfo.m_street == emo.GetStreet())
+      {
+        return NothingWasChanged;
+      }
+
+      // A feature was modified and equals to one in mwm (changes are rolled back).
+      if (sameAsInMWM)
+      {
+        // Feature was not yet uploaded. Since it's equal to one mwm we can remove changes.
+        if (editedFeatureInfo.m_uploadStatus != kUploaded)
+        {
+          RemoveFeatureFromStorageIfExists(fid.m_mwmId, fid.m_index);
+          // TODO(AlexZ): Synchronize Save call/make it on a separate thread.
+          Save(GetEditorFilePath());
+          Invalidate();
+          return SavedSuccessfully;
+        }
+      }
+
+      // If a feature is not the same as in mwm or it was uploaded
+      // we must save it and mark for an upcomming upload.
+    }
+    // A feature was NOT edited before and current changes are useless.
```

Comment:
Каждый кусок кода уникален, не нужно натягивать генерик правила на все случаи. Это не блокер, просто else смотрится криво, оторванно от ифа. Ну и пояснение перед return было бы более к месту.

> On Apr 7, 2016, at 14:28, mgsergio <notifications@github.com> wrote:
> 
> In indexer/osm_editor.cpp <https://github.com/mapsme/omim/pull/2778#discussion_r58856204>:
> 
> > +      {
> > +        // Feature was not yet uploaded. Since it's equal to one mwm we can remove changes.
> > +        if (editedFeatureInfo.m_uploadStatus != kUploaded)
> > +        {
> > +          RemoveFeatureFromStorageIfExists(fid.m_mwmId, fid.m_index);
> > +          // TODO(AlexZ): Synchronize Save call/make it on a separate thread.
> > +          Save(GetEditorFilePath());
> > +          Invalidate();
> > +          return SavedSuccessfully;
> > +        }
> > +      }
> > +
> > +      // If a feature is not the same as in mwm or it was uploaded
> > +      // we must save it and mark for an upcomming upload.
> > +    }
> > +    // A feature was NOT edited before and current changes are useless.
> У нас в коде пишут над, я делал по аналогии.
> 
> В прошлый раз, когда я писал внутри блока, ты предложил написать снаружи.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub <https://github.com/mapsme/omim/pull/2778/files/70e5e5d74135b19ee2eee212b16bd93922c97560#r58856204>